# Assessing subreddit origins via classification models.

Using CountVectorizer and Random Forest models we will assess which subreddit a sample of text originated from. 
The two sub-reddits we will be using are AskScience, and AskEngineers

### Start by import libraries

The first step will be to import the following libraries. These were determined during exploration and moved to the top for convention.

In [7]:
import requests
import pandas as pd
import numpy as np


from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

pd.set_option('display.max_columns', 999)

## Collecting data. 
Using PushShift API we will collect live subreddit data to build a model on.

#### AskEngineers Data

In [8]:
url = "https://api.pushshift.io/reddit/search/submission/?subreddit=AskEngineers&size=500&selftext:not='[removed]'&is_video=False"

req = requests.get(url)

engineers_df  = pd.DataFrame(req.json()['data'])

#### AskScience Data

In [9]:
url1 = "https://api.pushshift.io/reddit/search/submission/?subreddit=AskScience&size=500&selftext:not='[removed]'&is_video=False"

req1 = requests.get(url1)

science_df  = pd.DataFrame(req1.json()['data'])

## Concat Data

Next we will combine the two DataFrames via concatination and reset the index. (This will be necesarry later for merging data)

In [10]:
testing_df = pd.concat([science_df, engineers_df],sort=True)
testing_df = testing_df.reset_index().drop('index',axis=1)

## I set the Pandas Option to display more columns so I could explore the data better.

In [11]:
testing_df.head()

all_awardings allow_live_comments     author author_cakeday  \
0            []               False  [deleted]            NaN   
1            []               False  [deleted]            NaN   
2            []               False  [deleted]            NaN   
3            []               False  [deleted]            NaN   
4            []               False  [deleted]            NaN   

   author_created_utc author_flair_background_color author_flair_css_class  \
0                 NaN                                                 None   
1                 NaN                                                 None   
2                 NaN                                                 None   
3                 NaN                                                 None   
4                 NaN                                                 None   

  author_flair_richtext author_flair_template_id author_flair_text  \
0                   NaN                      NaN              None   
1                   NaN                      NaN              None   
2                   NaN                      NaN              None   
3                   NaN                      NaN              None   
4                   NaN                      NaN              None   

  author_flair_text_color author_flair_type author_fullname  \
0                    dark               NaN             NaN   
1                    dark               NaN             NaN   
2                    dark               NaN             NaN   
3                    dark               NaN             NaN   
4                    dark               NaN             NaN   

  author_patreon_flair author_premium awarders   banned_by  can_mod_post  \
0                  NaN            NaN       []  moderators         False   
1                  NaN            NaN       []  moderators         False   
2                  NaN            NaN       []  moderators         False   
3                  NaN            NaN       []  moderators         False   
4                  NaN            NaN       []  moderators         False   

   category  content_categories  contest_mode  created_utc crosspost_parent  \
0       NaN                 NaN         False   1580270334              NaN   
1       NaN                 NaN         False   1580269517              NaN   
2       NaN                 NaN         False   1580258983              NaN   
3       NaN                 NaN         False   1580257513              NaN   
4       NaN                 NaN         False   1580254317              NaN   

  crosspost_parent_list           domain  edited  \
0                   NaN  self.askscience     NaN   
1                   NaN  self.askscience     NaN   
2                   NaN  self.askscience     NaN   
3                   NaN  self.askscience     NaN   
4                   NaN  self.askscience     NaN   

                                           full_link  gilded gildings      id  \
0  https://www.reddit.com/r/askscience/comments/e...     NaN       {}  evhfdb   
1  https://www.reddit.com/r/askscience/comments/e...     NaN       {}  evh97v   
2  https://www.reddit.com/r/askscience/comments/e...     NaN       {}  evezk5   
3  https://www.reddit.com/r/askscience/comments/e...     NaN       {}  evempc   
4  https://www.reddit.com/r/askscience/comments/e...     NaN       {}  evduz5   

   is_crosspostable  is_meta  is_original_content  is_reddit_media_domain  \
0             False    False                False                   False   
1             False    False                False                   False   
2             False    False                False                   False   
3             False    False                False                   False   
4             False    False                False                   False   

   is_robot_indexable  is_self  is_video link_flair_background_color  \
0               False     True     False                     #ccccff   
1  

In [12]:
testing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 82 columns):
all_awardings                    933 non-null object
allow_live_comments              827 non-null object
author                           1000 non-null object
author_cakeday                   7 non-null object
author_created_utc               1 non-null float64
author_flair_background_color    428 non-null object
author_flair_css_class           63 non-null object
author_flair_richtext            705 non-null object
author_flair_template_id         22 non-null object
author_flair_text                153 non-null object
author_flair_text_color          448 non-null object
author_flair_type                705 non-null object
author_fullname                  705 non-null object
author_patreon_flair             705 non-null object
author_premium                   557 non-null object
awarders                         732 non-null object
banned_by                        293 non-null obje

## I did a partial cleaning
I set a threshhold on NaN in order to maintain the ```'selftext'``` column

In [13]:
part_cleaned_df = testing_df.dropna(axis=1, thresh=707)

In [14]:
part_cleaned_df.head()

all_awardings allow_live_comments     author awarders  can_mod_post  \
0            []               False  [deleted]       []         False   
1            []               False  [deleted]       []         False   
2            []               False  [deleted]       []         False   
3            []               False  [deleted]       []         False   
4            []               False  [deleted]       []         False   

   contest_mode  created_utc           domain  \
0         False   1580270334  self.askscience   
1         False   1580269517  self.askscience   
2         False   1580258983  self.askscience   
3         False   1580257513  self.askscience   
4         False   1580254317  self.askscience   

                                           full_link gildings      id  \
0  https://www.reddit.com/r/askscience/comments/e...       {}  evhfdb   
1  https://www.reddit.com/r/askscience/comments/e...       {}  evh97v   
2  https://www.reddit.com/r/askscience/comments/e...       {}  evezk5   
3  https://www.reddit.com/r/askscience/comments/e...       {}  evempc   
4  https://www.reddit.com/r/askscience/comments/e...       {}  evduz5   

   is_crosspostable  is_meta  is_original_content  is_reddit_media_domain  \
0             False    False                False                   False   
1             False    False                False                   False   
2             False    False                False                   False   
3             False    False                False                   False   
4             False    False                False                   False   

   is_robot_indexable  is_self  is_video link_flair_background_color  \
0               False     True     False                     #ccccff   
1               False     True     False                     #cc99ff   
2               False     True     False                     #aaddaa   
3               False     True     False                     #ccff99   
4               False     True     False                     #b4d8ff   

  link_flair_richtext link_flair_text_color link_flair_type  locked  \
0                  []                  dark            text   False   
1                  []                  dark            text   False   
2                  []                  dark            text   False   
3                  []                  dark            text   False   
4                  []                  dark            text   False   

   media_only  no_follow  num_comments  num_crossposts  over_18  \
0       False       True             2               0    False   
1       False       True             2               0    False   
2       False       True             2               0    False   
3       False       True             0               0    False   
4       False       True             0               0    False   

  parent_whitelist_status                                          permalink  \
0                 all_ads  /r/askscience/comments/evhfdb/i_dont_know_if_t...   
1                 all_ads        /r/askscience/comments/evh97v/proton_decay/   
2                 all_ads  /r/askscience/comments/evezk5/as_a_26_year_old...   
3                 all_ads  /r/askscience/comments/evempc/are_different_el...   
4                 all_ads  /r/askscience/comments/evduz5/how_does_the_atm...   

   pinned  pwls  retrieved_on  score selftext  send_replies  spoiler  \
0   False     6    1580271371      1      NaN          True    False   
1   False     6    1580270722      1      NaN          True    False   
2   False     6    1580262310      1      NaN          True    False   
3   False     6    1580261134      1      NaN          True    False   
4   False     6    1580258516      1      NaN          True    False   

   stickied   subreddit subreddit_id  subreddit_subscribers subreddit_type  \
0     False  askscience     t5_2qm4e               18550810         public   
1     False  askscience     t5_2q

In [15]:
part_cleaned_df.tail()

all_awardings allow_live_comments            author awarders  \
995            []               False   BiaxialObject48       []   
996            []               False     mixtape-maker       []   
997            []               False     denyingerrors       []   
998            []               False           Kyleh04       []   
999            []               False  The-Sober-Stoner       []   

     can_mod_post  contest_mode  created_utc             domain  \
995         False         False   1579402698  self.AskEngineers   
996         False         False   1579393189  self.AskEngineers   
997         False         False   1579391921  self.AskEngineers   
998         False         False   1579391914  self.AskEngineers   
999         False         False   1579391353  self.AskEngineers   

                                             full_link gildings      id  \
995  https://www.reddit.com/r/AskEngineers/comments...       {}  eqqtac   
996  https://www.reddit.com/r/AskEngineers/comments...       {}  eqoy7c   
997  https://www.reddit.com/r/AskEngineers/comments...       {}  eqoogs   
998  https://www.reddit.com/r/AskEngineers/comments...       {}  eqooe6   
999  https://www.reddit.com/r/AskEngineers/comments...       {}  eqojy3   

     is_crosspostable  is_meta  is_original_content  is_reddit_media_domain  \
995              True    False                False                   False   
996              True    False                False                   False   
997              True    False                False                   False   
998              True    False                False                   False   
999              True    False                False                   False   

     is_robot_indexable  is_self  is_video link_flair_background_color  \
995                True     True     False                               
996                True     True     False                               
997                True     True     False                               
998                True     True     False                               
999                True     True     False                               

    link_flair_richtext link_flair_text_color link_flair_type  locked  \
995                  []                  dark            text   False   
996                  []                  dark            text   False   
997                  []                  dark            text   False   
998                  []                  dark            text   False   
999                  []                  dark            text   False   

     media_only  no_follow  num_comments  num_crossposts  over_18  \
995       False       True             2               0    False   
996       False       True             2               0    False   
997       False       True             7               0    False   
998       False       True             4               0    False   
999       False       True             4               0    False   

    parent_whitelist_status  \
995                 all_ads   
996                 all_ads   
997                 all_ads   
998                 all_ads   
999                 all_ads   

                                             permalink  pinned  pwls  \
995  /r/AskEngineers/comments/eqqtac/rail_gun_with_...   False     6   
996  /r/AskEngineers/comments/eqoy7c/using_a_microc...   False     6   
997  /r/AskEngineers/comments/eqoogs/my_engineering...   False     6   
998  /r/AskEngineers/comments/eqooe6/is_a_raised_pl...   False     6   
999  /r/AskEngineers/comments/eqojy3/design_enginee...   False     6   

     retrieved_on  score                                           selftext  \
995    1579402699      1  Is it possible to make a rail gun where the pr...   
996    1579393191      1  The post title may sound a little confusing so...   
997    1579391922      1  I am currently a sophomore in aerospace engine...   
998    1579391915      1  I run

In [16]:
more_cleaned_df = part_cleaned_df.drop(['domain', 'author', 'wls', 'url', 'subreddit_type', 'subreddit_subscribers', 'subreddit_id', 'pwls', 'permalink', 'parent_whitelist_status', 'over_18', 'media_only', 'is_video', 'is_self', 'is_robot_indexable','gildings', 'full_link', 'contest_mode', 'can_mod_post', 'awarders', 'allow_live_comments', 'all_awardings', 'total_awards_received', 'retrieved_on'], axis=1)

## I removed more columns by hand
The columns were selected by "incriminating" data. I wanted to remove any data that automatically identified which subreddit the information came from. The only column maintained was ```'subreddit'``` which will be seperated to use a the ```'y_value'``` later.

In [17]:
more_cleaned_df.head()

created_utc      id  is_crosspostable  is_meta  is_original_content  \
0   1580270334  evhfdb             False    False                False   
1   1580269517  evh97v             False    False                False   
2   1580258983  evezk5             False    False                False   
3   1580257513  evempc             False    False                False   
4   1580254317  evduz5             False    False                False   

   is_reddit_media_domain link_flair_background_color link_flair_richtext  \
0                   False                     #ccccff                  []   
1                   False                     #cc99ff                  []   
2                   False                     #aaddaa                  []   
3                   False                     #ccff99                  []   
4                   False                     #b4d8ff                  []   

  link_flair_text_color link_flair_type  locked  no_follow  num_comments  \
0                  dark            text   False       True             2   
1                  dark            text   False       True             2   
2                  dark            text   False       True             2   
3                  dark            text   False       True             0   
4                  dark            text   False       True             0   

   num_crossposts  pinned  score selftext  send_replies  spoiler  stickied  \
0               0   False      1      NaN          True    False     False   
1               0   False      1      NaN          True    False     False   
2               0   False      1      NaN          True    False     False   
3               0   False      1      NaN          True    False     False   
4               0   False      1      NaN          True    False     False   

    subreddit thumbnail                                              title  \
0  askscience   default          I don’t know if this counts as one but...   
1  askscience   default                                       Proton Decay   
2  askscience   default  As a 26 year old, should I be worried about dy...   
3  askscience   default  Are different electroshock therapy treatments ...   
4  askscience   default                    How does the atmosphere...work?   

  whitelist_status  
0          all_ads  
1          all_ads  
2          all_ads  
3          all_ads  
4          all_ads

## Create X and y for fitting

In [18]:
X = more_cleaned_df.drop('subreddit', axis=1)
y = more_cleaned_df['subreddit']

## Instantiate Count Vectorizer for text analysis

In [84]:
cvec= CountVectorizer(min_df=3,ngram_range=(1,2))

## Replace np.NaN with empty strings as not to break CountVectorizer

In [20]:
X_clean = X.replace(np.nan, '', regex=True)

## Train Test Split

In order to validate model accuracy we need to seperate a testing class using train test split

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X_clean, y,
                                                   stratify=y,
                                                   random_state=42)

## Vectorize
Next we will fit our Count Vectorizer to the training data to generate word tokens for analysis

In [85]:
cvec.fit(X_train['selftext'])

X_vect_train = cvec.transform(X_train['selftext'])
X_vect_test = cvec.transform(X_test['selftext'])

## Build Token DataFrame

The Tokenized Data is generated as a sparse table. so we have to convert it a readable DataFrame. I'm assigning the column names back to the associated columns to explore trends later on.

In [86]:
vect_df_train = pd.DataFrame(columns=cvec.get_feature_names(), 
             data=X_vect_train.todense())
vect_df_test = pd.DataFrame(columns=cvec.get_feature_names(), 
             data=X_vect_test.todense())

## Merge Tokens

We have to merge the tokenized data back into our original DataFrame

In [87]:
Z_data_train = pd.merge(X_train.reset_index(drop=True), vect_df_train.reset_index(drop=True), left_index=True, right_index=True)
Z_data_test = pd.merge(X_test.reset_index(drop=True), vect_df_test.reset_index(drop=True), left_index=True, right_index=True)

## Filter for numerical Data

I then filter the data to return just numerical data

In [88]:
Z_train = Z_data_train.select_dtypes(['number'])
Z_test = Z_data_test.select_dtypes(['number'])

# Model Building
### Random Forest:

The first Model I'm testing is Random Forest

## Instantiate Random Forest Classifier

In [89]:
rf = RandomForestClassifier(n_jobs=6)

## Fit Model

In [90]:
rf.fit(Z_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=6,
                       oob_score=False, random_state=None, verbose=0,
                       warm_start=False)

## Score against Test Data

In [91]:
rf.score(Z_test, y_test)

0.948

## This feels like a really high scoring model so...

Instead of testing different models I will grid search over different estimators for Random Forest.

## Instantiate Grid Search 
With Range passed for n_estimator parameters

In [29]:
params = {
    'n_estimators' : list(range(2,25))
}

grid = GridSearchCV(RandomForestClassifier(n_jobs=-1), param_grid=params)

## Fit Gridsearch to Training Data

In [30]:
grid.fit(Z_train, y_train)

GridSearchCV(cv=None, error_score=nan,
             estimator=RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                              class_weight=None,
                                              criterion='gini', max_depth=None,
                                              max_features='auto',
                                              max_leaf_nodes=None,
                                              max_samples=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              n_estimators=100, n_jobs=-1,
                                              oob_score=False,
                                              rand

## Score Best Estimator against Train and Test

In [31]:
print(grid.score(Z_train, y_train))
print(grid.score(Z_test, y_test))

0.9946666666666667
0.936


## Save Model for Later Use

In [34]:
best_model = grid.best_estimator_

### SVM:

In [92]:
from sklearn.svm import SVC

In [118]:
grid_svc = GridSearchCV(SVC(break_ties=True, random_state=42),
                   param_grid={'C':np.logspace(-2,15,num=50)},
                   n_jobs=8)

# Fit on training data.
grid_svc.fit(Z_train, y_train)
# Evaluate model.
grid_svc.score(Z_test, y_test)

0.984

In [95]:
grid_svc.best_estimator_

SVC(C=17575.10624854793, break_ties=True, cache_size=200, class_weight=None,
    coef0=0.0, decision_function_shape='ovr', degree=3, gamma='scale',
    kernel='rbf', max_iter=-1, probability=False, random_state=42,
    shrinking=True, tol=0.001, verbose=False)

### Experimenting with XGBoost

In [119]:
import xgboost as xgb

In [120]:
xb = xgb.XGBClassifier()

In [122]:
xb.fit(Z_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [123]:
xb.score(Z_test,y_test)

0.996

In [126]:
xg_rf = xgb.XGBRFClassifier()
params = {
    'n_estimators' : list(range(2,25))
}

grid_xg_rf = GridSearchCV(xgb.XGBRFClassifier(n_jobs=11), param_grid=params)
grid_xg_rf.fit(Z_train, y_train)
grid_xg_rf.score(Z_test, y_test)

0.984

# Conclusion

Using Subreddit Data we were able to build a model that can sort between Ask Science and Ask Engineers with 93.6% accuracy using Random Forest and 98.4 using a Support Vector Machine.

*These numbers may change when re-running cells due to new data but, they will likely be similar*

# Exploring Trends in Data



In [96]:
train_total = pd.merge(left=Z_train.reset_index(drop=True), right=y_train.reset_index(drop=True),
                      left_index=True, right_index=True)

In [97]:
test_total = pd.merge(left=Z_test.reset_index(drop=True), right=y_test.reset_index(drop=True),
                      left_index=True, right_index=True)

In [98]:
total_df = pd.concat([train_total, test_total], sort=True)
total_df = total_df.reset_index().drop('index',axis=1)

In [99]:
cleaned_analysis = total_df.drop('level_0', axis=1)

In [100]:
cleaned_analysis.head()

00  000  01  07  10  100  10am  11  11am  12  12pm  12pm et  13  14  15  \
0   0    0   0   0   0    0     0   0     0   0     0        0   0   0   0   
1   0    0   0   0   0    0     0   0     0   0     0        0   0   0   0   
2   0    0   0   0   0    0     0   0     0   0     0        0   0   0   0   
3   0    0   0   0   0    0     0   0     0   0     0        0   0   0   0   
4   0    0   0   0   0    0     0   0     0   0     0        0   0   0   0   

   15 ut  16  16 ut  17  17 19  17 ut  18  18 ut  19  19 ut  1pm  1pm et  1st  \
0      0   0      0   0      0      0   0      0   0      0    0       0    0   
1      0   0      0   0      0      0   0      0   0      0    0       0    0   
2      0   0      0   0      0      0   0      0   0      0    0       0    0   
3      0   0      0   0      0      0   0      0   0      0    0       0    0   
4      0   0      0   0      0      0   0      0   0      0    0       0    0   

   20  20 years  2010  2014  2016  2017  2018  2019  2020  2021  21  23  24  \
0   0         0     0     0     0     0     0     0     0     0   0   0   0   
1   0         0     0     0     0     0     0     0     0     0   0   0   0   
2   0         0     0     0     0     0     0     0     0     0   0   0   0   
3   0         0     0     0     0     0     0     0     0     0   0   1   0   
4   0         0     0     0     0     0     0     0     0     0   0   0   0   

   25  250  29  2d  2pm  2pm et  30  3d  3d model  3pm  3pm et  40  50  500  \
0   0    0   0   0    0       0   0   0         0    0       0   0   0    0   
1   0    0   0   0    0       0   0   0         0    0       0   0   0    0   
2   0    0   0   0    0       0   0   0         0    0       0   0   0    0   
3   0    0   0   0    0       0   0   0         0    0       0   0   0    0   
4   0    0   0   0    0       0   0   0         0    0       0   0   0    0   

   60  90  ability  able  able to  about  about an  about how  about it  \
0   0   0        0     0        0      0         0          0         0   
1   0   0        0     0        0      0         0          0         0   
2   0   0        0     0        0      0         0          0         0   
3   0   0        0     0        0      0         0          0         0   
4   0   0        0     0        0      1         0          0         0   

   about me  about my  about one  about the  about this  about what  \
0         0         0          0          0           0           0   
1         0         0          0          0           0           0   
2         0         0          0          0           0           0   
3         0         0          0          0           0           0   
4         0         0          0          0           0           0   

   about your  above  absolutely  academic  acceleration  acceptable  \
0           0      0           0         0             0           0   
1           0      0           0         0             0           0   
2           0      0           0         0             0           0   
3           0      0           0         0             0           0   
4           0      0           0         0             0           0   

   accepted  accepted an  access  access to  accessible  accomplish  \
0         0            0       0          0           0           0   
1         0            0       0          0           0           0   
2         0            0       0          0           0           0   
3         0            0       0          0           0           0   
4         0            0       0          0           0           0   

   accomplish this  according  according to  accurate  achieve  achieve the  \
0                0          0             0         0        0            0   
1                0          0             0         0        0            0   
2                0          0             0         0        0            0   
3                0          0             0

In [43]:
total_df.shape

(1000, 2028)

In [101]:
science_df2 = total_df[total_df['subreddit_y']=='askscience']

In [102]:
engineer_df2 = total_df[total_df['subreddit_y']=='AskEngineers']

In [103]:
science_df2.reset_index(inplace=True)

In [104]:
engineer_df2.reset_index(inplace=True)

In [105]:
engineer_clean = engineer_df2.drop(['subreddit_y','level_0'], axis=1)

In [106]:
science_clean = science_df2.drop(['subreddit_y','level_0'], axis=1)

In [107]:
science_clean.sum().sort_values(ascending=False).head(40)

created_utc     783832367939
index                 249702
score                 229958
num_comments           29112
the                     1020
and                      706
of                       629
to                       506
in                       423
is                       266
that                     236
for                      210
at                       205
on                       201
we                       190
https                    188
it                       163
are                      150
be                       132
with                     130
research                 121
about                    120
from                     119
of the                   117
www                      115
as                       113
this                     110
or                       109
in the                   105
com                      104
my                       101
you                      100
can                       96
https www                 94
our           

In [108]:
engineer_clean.sum().sort_values(ascending=False).head(40)

created_utc     789929505839
index                 249798
num_comments            5159
the                     1955
to                      1729
and                     1413
in                      1043
of                       972
is                       771
my                       756
that                     712
for                      685
it                       637
score                    506
have                     488
this                     446
be                       434
with                     429
but                      428
on                       392
would                    349
as                       323
or                       312
am                       312
at                       307
what                     303
an                       300
you                      299
if                       290
engineering              283
not                      278
can                      277
so                       269
like                     267
are           

In [109]:
engineer_only = engineer_clean.sum()[science_clean.sum()==0].sort_values(ascending=False)

In [110]:
science_only = science_clean.sum()[engineer_clean.sum()==0].sort_values(ascending=False)

In [111]:
engineer_only.shape

(1283,)

In [112]:
science_only.shape

(301,)

In [113]:
total_df.shape

(1000, 4370)

In [115]:
science_only.head(40)

ut                53
et                52
twitter           37
edu               36
nih               36
cells             35
ama               29
ph                26
ut ask            26
nih gov           25
your questions    24
twitter com       24
https twitter     24
ask us            22
us anything       21
we ll             18
scientists        18
institute         18
scientific        18
me anything       17
biological        17
published         17
answer your       16
16 ut             16
neutrons          15
19                15
disease           14
black             14
00                14
species           13
mission           13
medicine          13
the brain         13
population        13
children          13
emotional         12
dna               12
stars             12
et 16             12
articles          11
dtype: int64

In [117]:
engineer_only.head(40)

mechanical                68
advice                    57
the company               43
internship                42
interview                 40
thank you                 39
thank                     39
guys                      36
my first                  32
to work                   32
mechanical engineering    30
any advice                30
offer                     29
graduated                 27
am currently              27
my career                 27
to start                  26
of engineering            25
boss                      25
positions                 25
an engineer               24
in engineering            24
path                      23
it was                    23
motor                     23
you guys                  23
ee                        22
sector                    22
asked                     21
my resume                 21
interviews                21
civil                     21
is my                     21
my job                    20
if anyone     

In [ ]:
plot.